### PR TITLE
HifiGAN MelSpectrogram Vocoder Model

### PR DESCRIPTION
**Pull request by felixkreuk**
This PR introduces the HifiGAN vocoder to NeMo.

### CODE REVIEW

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  accelerator: ddp
+  accumulate_grad_batches: 1
+  checkpoint_callback: False  # Provided by exp_manager
+  logger: False  # Provided by exp_manager
+  log_every_n_steps: 100
+  check_val_every_n_epoch: 10
+  automatic_optimization: false
+
+exp_manager:
+  exp_dir: /results
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:80]: Can you remove this line?

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  name: ${name}
+  create_tensorboard_logger: false
+  create_checkpoint_callback: True
+
+  version: ???
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:85]: Does this need to be `???`?

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  use_datetime_version: false
+  resume_if_exists: true
+  resume_ignore_no_checkpoint: true
+  resume_past_end: true
+  
+  create_wandb_logger: true
+  wandb_logger_kwargs:
+    project: hifi
+    name: ${exp_manager.version}
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:94]: These seem to be your experimental config, can we reset them to our usual defaults?

File: nemo/collections/tts/models/hifigan.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+        # train discriminator
+        self.optim_d.zero_grad()
+        mpd_score_real, mpd_score_gen, _, _ = self.mpd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_mpd, _, _ = discriminator_loss(mpd_score_real, mpd_score_gen)
+        msd_score_real, msd_score_gen, _, _ = self.msd(y=audio, y_hat=audio_pred.detach())
+        loss_disc_msd, _, _ = discriminator_loss(msd_score_real, msd_score_gen)
+        loss_d = loss_disc_msd + loss_disc_mpd
+        self.manual_backward(loss_d, self.optim_d)
+        self.optim_d.step()
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:104]: Is there any reason, you choose to do the manual backwards as opposed to relying on lightning (like in melGAN)?

File: nemo/collections/tts/modules/hifigan_modules.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (blisc) [inline nemo/collections/tts/modules/hifigan_modules.py:1]: Blocking

File: nemo/collections/tts/models/hifigan.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
+        if isinstance(cfg, dict):
+            cfg = OmegaConf.create(cfg)
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        self.audio_to_melspec_precessor = instantiate(cfg.preprocessor)
+        # use a different melspec extractor because:
+        # 1. we need to pass grads
+        # 2. we need remove fmax limitation
+        self.trg_melspec_fn = instantiate(cfg.preprocessor, highfreq=None, use_grads=True)
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:43]: Do you think that this is better than defining a new audio -> mel module?

File: nemo/collections/tts/models/hifigan.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+    def validation_step(self, batch, batch_idx):
+        audio, audio_len = batch
+        audio_mel, audio_mel_len = self.audio_to_melspec_precessor(audio, audio_len)
+        audio_pred = self(spec=audio_mel)
+
+        audio_pred_mel, _ = self.audio_to_melspec_precessor(audio_pred.squeeze(1), audio_len)
+        loss_mel = F.l1_loss(audio_mel, audio_pred_mel)
+
+        self.log("val_loss", loss_mel, prog_bar=True, sync_dist=True)
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:142]: Doesn't this just log the loss at one step? As opposed to the validation loss over the entire val dataset.

File: nemo/collections/tts/models/hifigan.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+
+def generator_loss(disc_outputs):
+    loss = 0
+    gen_losses = []
+    for dg in disc_outputs:
+        l = torch.mean((1 - dg) ** 2)
+        gen_losses.append(l)
+        loss += l
+
+    return loss, gen_losses
```
Comment (blisc) [inline nemo/collections/tts/models/hifigan.py:238]: I think we currently require these to be children of the NeMo Loss class

**Review by blisc — CHANGES_REQUESTED** [review]

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 1 above)
Comment (felixkreuk) [reply examples/tts/conf/hifigan/hifigan.yaml:80]: which line? i see lines 77-80 in the commment

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 2 above)
Comment (felixkreuk) [reply examples/tts/conf/hifigan/hifigan.yaml:85]: see below

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 3 above)
Comment (felixkreuk) [reply examples/tts/conf/hifigan/hifigan.yaml:94]: ok, i will reset exp_manager section to default params (like melgan)

File: nemo/collections/tts/models/hifigan.py (hunk 4 above)
Comment (felixkreuk) [reply nemo/collections/tts/models/hifigan.py:104]: It is more efficient that way, 1 generator forward pass for bath updates (G and D). as opposed to 1 generator forward pass for 1 generator update, and another generator pass for 1 discriminator update. this is also more in line with the optimization described in the original source code.

File: nemo/collections/tts/models/hifigan.py (hunk 6 above)
Comment (felixkreuk) [reply nemo/collections/tts/models/hifigan.py:43]: not sure i follow.

File: nemo/collections/tts/models/hifigan.py (hunk 7 above)
Comment (felixkreuk) [reply nemo/collections/tts/models/hifigan.py:142]: no, according to PTL, when you log in `val_step` the default behavior is to aggregate all steps into an epoch summary of the metric. see [here](https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#log).

File: nemo/collections/tts/models/hifigan.py (hunk 8 above)
Comment (felixkreuk) [reply nemo/collections/tts/models/hifigan.py:238]: done

File: examples/tts/conf/hifigan/hifigan.yaml [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      shuffle: true
+      batch_size: 16
+      num_workers: 4
+
+  validation_ds:
+    dataset:
+      _target_: "nemo.collections.tts.data.datalayers.AudioDataset"
+      manifest_filepath: ${validation_datasets}
+      max_duration: null
+      min_duration: 0.75
```
Comment (blisc) [inline examples/tts/conf/hifigan/hifigan.yaml:28]: Let's set min_duration to null for validation

File: nemo/collections/tts/models/hifigan.py (hunk 6 above)
Comment (blisc) [reply nemo/collections/tts/models/hifigan.py:44]: Rather than using the FilterBankFeatures modules, wouldn't it be faster to create a smaller more efficient module?

File: examples/tts/conf/hifigan/hifigan.yaml (hunk 9 above)
Comment (felixkreuk) [reply examples/tts/conf/hifigan/hifigan.yaml:28]: ok, i took this config from melgan [here](https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/melgan.yaml#L31). i will fix that

Comment (blisc) [reply examples/tts/conf/hifigan/hifigan.yaml:28]: Yea melGAN properly also needs this fix then.

Comment (felixkreuk) [reply examples/tts/conf/hifigan/hifigan.yaml:28]: no problem

File: nemo/collections/tts/models/hifigan.py (hunk 6 above)
Comment (felixkreuk) [reply nemo/collections/tts/models/hifigan.py:44]: resolved via slack

**Comment by okuchaiev** [comment]
please check DCO's bot output and sign unsigned commits

**Comment by felixkreuk** [comment]
> please check DCO's bot output and sign unsigned commits

done

**Review by blisc — APPROVED** [review]
